### PR TITLE
Call fstat only once per refresh

### PR DESCRIPTION
**Pull request by hoovercj**
This drastically improves performance with SD on windows. it takes the refresh time down from ~20 seconds to ~5 seconds for a changelist with ~20 files.

Please verify that it still behaves as expected with perforce and with linux or osx. I think I'm handling line endings correctly but you never can be too sure :-)

### CODE REVIEW

**Comment by hoovercj** [comment]
Addresses #51

File: src/scm/Model.ts [hunk 1]
```diff
@@ -347,6 +320,43 @@ export class Model implements Disposable {
...
+            if (matches) {
+                files.push(matches[1]);
+            }
+        });
+
+        return files;
+    }
 
+    private async getFstatInfoForFiles(files: string[]): Promise<any> {
+        const fstatOutput: string = await Utils.getOutput(`fstat ${files.join(' ')}`);
```
Comment (stef-levesque) [inline src/scm/Model.ts]: This won't work for files with spaces in the path
(e.g. //depot/source/file with space.txt#1)

Comment (stef-levesque) [inline src/scm/Model.ts]: Each file path will need to be wrapped with quotes (")

Comment (stef-levesque) [inline src/scm/Model.ts]: `fstat "${files.join('" "')}"` should do

File: src/scm/Model.ts [hunk 2]
```diff
@@ -347,6 +320,43 @@ export class Model implements Disposable {
...
+        // Windows will have lines end with \n\r.
+        // Each file has multiple lines of output separated by a blank line.
+        // Splitting on \n\r?\n will separate the output for each file.
+        const fstatFiles = fstatOutput.trim().split(/\n\r?\n/);
```
Comment (stef-levesque) [inline src/scm/Model.ts]: isn't it rather \r\n ?

Comment (stef-levesque) [inline src/scm/Model.ts]: To be compatible with both `\n` and `\r\n`,
I think it should be: `const fstatFiles = fstatOutput.trim().split(/\r?\n/);`

Comment (stef-levesque) [inline src/scm/Model.ts]: Oops... you mean splitting file sections with an empty line... got it. 😅

File: src/scm/Model.ts [hunk 3]
```diff
@@ -347,6 +320,43 @@ export class Model implements Disposable {
...
+        return fstatFiles.map((file) => {
+            const lines = file.split('\n');
+            const lineMap = {};
+            lines.forEach(line => {
+                // ... Name Value
+                const matches = line.match(/[.]{3} (\w+) (.+)/);
```
Comment (stef-levesque) [inline src/scm/Model.ts]: This won't work with `... isMapped`
There is no 2nd match.
Although, not a blocker.

Comment (stef-levesque) [inline src/scm/Model.ts]: This should work:
```typescript
const matches = line.match(/[.]{3} (\w+)\s?(.+)?/);
if (matches) {
    lineMap[matches[1]] = matches[2] ? matches[2] : 'true';
}
```

Comment (hoovercj) [inline src/scm/Model.ts]: Is mapped used? Or are you just thinking about total correctness?

File: src/scm/Model.ts (hunk 1 above)
Comment (hoovercj) [inline src/scm/Model.ts]: Yep, that works.

File: src/scm/Model.ts (hunk 2 above)
Comment (hoovercj) [inline src/scm/Model.ts]: I've fixed my mistake in the comment and edited the comment a bit to explain why I think this is correct.

File: src/scm/Model.ts (hunk 3 above)
Comment (stef-levesque) [inline src/scm/Model.ts]: It's not being used right now, but eventually I hope to save all fields, and contextualize commands if the file is mapped, shelved, etc...

see [P4 fstat](https://www.perforce.com/perforce/r14.2/manuals/cmdref/p4_fstat.html) for more info

File: src/scm/Model.ts [hunk 4]
```diff
@@ -341,19 +341,22 @@ export class Model implements Disposable {
...
+        // which will split the output for each file.
         const fstatFiles = fstatOutput.trim().split(/\n\r?\n/);
         return fstatFiles.map((file) => {
             const lines = file.split('\n');
             const lineMap = {};
             lines.forEach(line => {
-                // ... Name Value
-                const matches = line.match(/[.]{3} (\w+) (.+)/);
+                // ... Key Value
+                const matches = line.match(/[.]{3} (\w+) (.+)?/);
```
Comment (stef-levesque) [inline src/scm/Model.ts]: The space also need to be optional
```regex
/[.]{3} (\w+)\s?(.+)?/
```

File: src/scm/Model.ts [hunk 5]
```diff
@@ -296,7 +296,7 @@ export class Model implements Disposable {
             const uri = Uri.file(clientFile);
             const resource: Resource = new Resource(uri, change, action);
 
-            if (change === 'default change') {
+            if (change.startsWith('default')) {
```
Comment (hoovercj) [inline src/scm/Model.ts]: I noticed that source depot actually outputs "... change default" with fstat (without ztag). If perforce really is outputting "default change", even without ztag, this should still catch it, but please make sure files in the default change show up.

File: src/scm/Model.ts (hunk 3 above)
Comment (hoovercj) [inline src/scm/Model.ts]: Nice. No reason to leave a landmine around then :-)

File: src/scm/Model.ts (hunk 4 above)
Comment (hoovercj) [inline src/scm/Model.ts]: That's what I get for not just copying and pasting. "no no, i can type it just fine" :P

Comment (hoovercj) [inline src/scm/Model.ts]: The space shouldn't be captured as part of the capture group, though, so I'll use this: `[/[.]{3} (\w+)[ ]*(.+)?/`

Does that look ok?

File: src/scm/Model.ts (hunk 5 above)
Comment (stef-levesque) [inline src/scm/Model.ts]: `p4 opened` will output `default change`
but `p4 fstat` will output `... change default`

Don't ask me why 😞 

So, yes, your change is valid.

**Comment by stef-levesque** [comment]
Thanks a lot! This PR should make it more efficient and more robust. 💪

**Pull request merged**